### DISCUSSION TITLE
BlenderBim : Storing Pset fields on current scene rather than individual objects

### DISCUSSION THREAD

**Discussion by Gorgious56** [IfcOpenShell/IfcOpenShell#1691] [Ideas]
I'm working on editing object operators to execute on all selected objects rather than only the active one. 
See https://github.com/Gorgious56/IfcOpenShell/tree/apply-object-operators-to-all-selected-objects

I've hit a roadblock on some operators, because some properties are stored on object-level so the result may not be the same as expected. 

Illustration :

![BSE_40](https://user-images.githubusercontent.com/25156105/130756730-9aff90fe-ba28-4d66-9352-5e4b6b6c02aa.gif)

The operator will correctly execute and add a Pset, but since the enum property is changed only on the active object, the actual Pset that's added on the other selected objects is not the expected one.

Q : 
Shouldn't the object (and material) `PsetProperties` be stored on the current scene rather than on every object ? 
In most cases only one instance of this enum will be displayed to the user so I think it makes sense to store it on the scene object, especially since it's already the case for Psets for Task, Resource, Profile or WorkSchedule. This would also help uncouple the logic from the blender API a bit more.

**Comment by Moult** [comment]
The pset properties will change depending on the selected object. The properties you can add depend on the object's IFC class. Therefore, implementing "add pset to all selected" should be done very carefully, since you may add an invalid pset to an object.

**Reply by Gorgious56** [reply]
Alright I understand. Do you think there is a need to try and refactor the code to do it ? It may be more trouble than it's worth then. I was thinking silently passing if the object's type and IFC type don't match the active object but that may not be desirable or easily doable.

**Reply by Moult** [reply]
Maybe skip it for now :)

**Reply by vulevukusej** [reply]
@Gorgious56  - Is this idea dead or do you still plan on revisiting this some day?

**Reply by Gorgious56** [reply]
No I don't think it's dead, although as Dion pointed out, it's not as straightforward as iterating over all selected objects and adding the property set since there can be intricate dependencies prenventing some associations. I think it will be solved along other QOL features once everything is more mature and stable :)

**Reply by vulevukusej** [reply]
Yeah you´re absolutely right, I suppose other software get away with it because they give the user the freedom to do completely edit the properties without any consideration of the ifc standards.  I agree with Dion's strict implementation here though so I'm happy to wait 👍

**Reply by vulevukusej** [reply]
Hey @Gorgious56 - not content to just sit on this, I decided to give it a crack ;).   I created my own version of the "bim.add_pset" class but this time added the ability to add the selected pset to all selected objects.   When looping through each object, I checked whether or not the  type is included in the "ApplicableEntity" attribute of the chosen pset.   So far so good, it even gives the user a small error message if they try to do this.

Are there any other checks that would need to be done? @Moult would this satisfy you as well?

```
class BulkAddPsets(bpy.types.Operator):
    bl_label = "Add Pset to selected objects"
    bl_idname = "ifc_mapping.add_pset_to_selected"
    bl_options = {"REGISTER", "UNDO"}
    obj: bpy.props.StringProperty()
    obj_type: bpy.props.StringProperty()
    pset: bpy.props.StringProperty()
    

    def getApplicableEntities(self, context):
        templates = IfcStore.pset_template_file.by_type("IfcPropertySetTemplate")

        for template in templates:
            pset_name = template.Name

            if pset_name == self.pset:
                return template.ApplicableEntity.split(",")

    def execute(self, context):
        #print(getPsetTemplates(self,context))
        return IfcStore.execute_ifc_operator(self, context)
        
    def _execute(self, context):
        self.file = IfcStore.get_file()
        selected_objects = context.selected_objects
        applicable_entities = self.getApplicableEntities(context)


        for object in selected_objects:
            props = object.BIMObjectProperties
            try:
                data = ObjectData.Data.products[props.ifc_definition_id]
            except:
                print(f"IFC Element:{props.ifc_definition_id} not found")
                continue
            object_ifc_class = data["type"]            
            ifc_definition_id = object.BIMObjectProperties.ifc_definition_id

            if object_ifc_class in applicable_entities:
                run(
                "pset.add_pset",
                self.file,
                **{
                    "product": self.file.by_id(ifc_definition_id),
                    "name": self.pset,
                }
                )
                Data.load(IfcStore.get_file(), ifc_definition_id)
            else:
                print(f"{self.pset} was not added to to object:{ifc_definition_id} since {object_ifc_class} is not an applicable entity for this Pset")

        self.report({'INFO'}, 'Finished applying changes')
        return {"FINISHED"}`
```

**Reply by vulevukusej** [reply]
End product:

![image](https://user-images.githubusercontent.com/83825269/137149131-22cd387c-b231-424e-874d-c80d156c8a69.png)

**Reply by vulevukusej** [reply]
To get all properties in a template, and not just those that are stored on the object I edited my panel so:



    def draw(self, context):
        props = context.scene.BIMPsetTemplateProperties
        layout = self.layout
        layout.prop(props, "pset_template_files", text="")
        layout.prop(props, "pset_templates", text="", icon="COPY_ID")
        
        row = self.layout.row(align=True)
        op = row.operator("ifc_mapping.add_pset_to_selected", icon="ADD")
        op.pset = self.getPsetName(context, props.pset_templates)
        
    def getPsetName(self, context, pset_id):
        pset_enum = [pset_tuple for pset_tuple in getPsetTemplates(self,context)]
        for pset in pset_enum:
            if pset[0] == pset_id:
                return pset[1]

**Reply by Moult** [reply]
Sorry I forgot to reply. I'll check this out when I get to a computer. The approach described sounds correct, but at face value I wonder why a new ui panel is required. Can't adding a pset just occur to all selected objects, then we can also have a new copy pset button?

The value of the "rename" panel seems in the ability to map psets - that is indeed a useful function and can be kept. But the second subpanel "add psets to selected" probably doesn't need to exist from a UX perspective.

**Reply by Moult** [reply]
Would it be possible to submit a PR for this so we can review and track it?

**Reply by vulevukusej** [reply]
> Sorry I forgot to reply. I'll check this out when I get to a computer. The approach described sounds correct, but at face value I wonder why a new ui panel is required. Can't adding a pset just occur to all selected objects, then we can also have a new copy pset button?
> 
> The value of the "rename" panel seems in the ability to map psets - that is indeed a useful function and can be kept. But the second subpanel "add psets to selected" probably doesn't need to exist from a UX perspective.

@Moult  - Now I remember another reason why I created it as a separate add-on.  Normally when you select an object and then choose a pset from the drop-down list you can only see psets that are applicable to the last selected object.  In my version, the dropdown shows _all_ psets that are available in the template.  Whether or not they get added in the end depends on the applicable entities attribute.  Do you have a preference for how to go about this?

**Reply by Moult** [reply]
All psets that are available in the template are huge. I count 420 psets available in IFC. Am I interpreting you correctly?

**Reply by vulevukusej** [reply]
No sorry I was referring to custom-defined templates, not those that are natively available in IFC.

**Reply by vulevukusej** [reply]
Having said that though, you've made me wonder whether I'm just allowing the user to be excessively lazy, ie. just bulk adding a particular pset to all objects instead of purposefully, say, adding a specific pset for masonry wall types to only masonry walls.

**Reply by vulevukusej** [reply]
On the flipside though, if i click on two different element types, lets say i click on a few walls and on the last click i accidentally click on a window - the user will only see psets that are strictly applicable to IfcWindow's for example.  Somehow I feel as if that will be a little confusing/annoying for the user..

**Reply by Moult** [reply]
Ah OK. I think from a UX perspective it is odd to separate standard and custom psets. Either today or tomorrow I'll be coding some bulk property copying features which I need for a commercial job, so I might copy parts of your code and see what happens.

**Comment by Gorgious56** [comment]
Good stuff !! 👍  I won't lie, I'm not so well versed in the Pset area yet so I won't be able to comment on the functionality but I can give a few pointers (Of course @Moult has the final say in these matters, but he's a little busy on a mega-refactor atm ^^)
-  Try to use snake_case and not camelCase for function names (some functions in existing code do use camelCase but they're incrementally being renamed, your work could be an opportunity to do it)
- I think you could simplify `getPsetName` and use something like `op.pset = next(template[1] for template in getPsetTemplates(selft, context) if template[0] == props.pset_templates)` (if I understood correctly which may not be the case :p )
- BlenderBim operators' `bl_idname` follow the convention `bim.what_the_operator_does`

Other than that, if you're willing to share, maybe you could submit a Pull Request so we can test it out ?

Cheers !

**Reply by vulevukusej** [reply]
Yeah I'll post it when I get to my pc again.  It was a seperate panel since it was originally meant to be a stand-alone plugin for my own purposes but I'll happily integrate it into the existing BB panel.
